File: peer_lib/sig/arkoala/framework/java/src/org/koalaui/arkoala/PeerNode.java
/*
 * Copyright (c) 2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.koalaui.arkoala;

import java.util.Map;

public class PeerNode extends IncrementalNode {
    private static final int PEER_NODE_TYPE = 11;
    public NativePeerNode peer;
    public PeerNode(ArkUINodeType type, ComponentBase component, int flags) {
        super(PEER_NODE_TYPE);
        int id = 0; // TODO: use id
        long ptr = NativeModule._CreateNode(type.value, id, flags);
        this.peer = new NativePeerNode(ptr, NativeModule._GetNodeFinalizer());
    }
    void applyAttributes(Map<String, Object> attrs) {}
}

File: arkgen/tests/unit/setup/kotlin/sig/arkoala-kotlin/framework/src/PeerNode.kt
/*
 * Copyright (c) 2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package koalaui.arkoala

import kotlin.collections.ArrayList
import kotlin.collections.HashMap
import koalaui.interop.pointer

import arkui.component.ArkRootPeer

// add inheritance from NativePeerNode
public open class PeerNode(peerPtr: pointer, val id: Int, val name: String, flags: Int) {
    public val peer: NativePeerNode

    init {
        peer = NativePeerNode.create(peerPtr)
        peerNodeRawTail.add(this)
    }

    companion object {
        protected var currentId: Int = 1000
        private val PEER_NODE_TYPE: UInt = 11u

        fun nextId(): Int {
            currentId += 1
            return currentId
        }

        fun generateRootPeer(): PeerNode {
            return ArkRootPeer.create(null)
        }

        private val peerNodeMap = HashMap<Int, PeerNode>()
        private val peerNodeRawTail = ArrayList<PeerNode>()

        private fun updatePeerNodeMap() {
            if (peerNodeRawTail.isEmpty()) {
                return
            }
            peerNodeRawTail.forEach({ peer ->
                peerNodeMap.set(peer.id, peer)
            })
            peerNodeRawTail.clear()
        }

        fun findPeerByNativeId(id: Int): PeerNode? {
            updatePeerNodeMap()
            return peerNodeMap.get(id)
        }
    }

    fun getPeerPtr(): pointer {
        return peer.ptr
    }

    fun applyAttributes(attrs: HashMap<String, Any>) {}
}